### PR TITLE
Allow marking quickstarts as favorite

### PR DESCRIPTION
**Pull request by victoria-dos**
[Jira issue](https://issues.redhat.com/browse/RHCLOUD-29274)

### CODE REVIEW

File: cmd/favorite/favorite.go
Comment (Hyperkid123) [inline cmd/favorite/favorite.go:1]: It should be clear this script is for testing only.

File: cmd/migrate/migrate.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ func main() {
 	godotenv.Load()
 	config.Init()
 	database.Init()
-	err := database.DB.AutoMigrate(&models.Quickstart{}, &models.QuickstartProgress{}, &models.Tag{}, &models.HelpTopic{})
+	err := database.DB.AutoMigrate(&models.Quickstart{}, &models.QuickstartProgress{}, &models.Tag{}, &models.HelpTopic{}, &models.FavoriteQuickstart{})
```
Comment (Hyperkid123) [inline cmd/migrate/migrate.go:15]: I would not add this here until we have the actual migration figured out.

File: main.go [hunk 2]
```diff
@@ -66,12 +66,14 @@ func main() {
...
 		sub.Route("/progress", routes.MakeQuickstartsProgressRouter)
 		sub.Route("/helptopics", routes.MakeHelpTopicsRouter)
+		sub.Route("/favorites", routes.MakeFavoriteQuickstartsRouter)
 		sub.Handle("/spec/*", http.StripPrefix("/api/quickstarts/v1/spec", fs))
 	})
 	mr.Get("/", probe)
 	mr.Handle("/metrics", promhttp.Handler())
 	r.Get("/test", probe)
 
+	// fmt.Println(cfg.ServerAddr)
```
Comment (Hyperkid123) [inline main.go:76]: This can be removed

File: main.go [hunk 3]
```diff
@@ -66,12 +66,14 @@ func main() {
 		sub.Route("/quickstarts", routes.MakeQuickstartsRouter)
 		sub.Route("/progress", routes.MakeQuickstartsProgressRouter)
 		sub.Route("/helptopics", routes.MakeHelpTopicsRouter)
+		sub.Route("/favorites", routes.MakeFavoriteQuickstartsRouter)
```
Comment (Hyperkid123) [inline main.go:69]: I would hide this behind an env variable for now until the full workflow is finished.

File: pkg/database/db.go [hunk 4]
```diff
@@ -42,10 +42,13 @@ func Init() {
 	if !DB.Migrator().HasTable(&models.HelpTopic{}) {
 		DB.Migrator().CreateTable(&models.HelpTopic{})
 	}
+	if !DB.Migrator().HasTable(&models.FavoriteQuickstart{}) {
```
Comment (Hyperkid123) [inline pkg/database/db.go:45]: Again, hide this until all features are done.

File: pkg/routes/favorite_quickstart.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+	var favQuickstart, err = SwitchFavorite(accountId, payload.QuickstartName, payload.Favorite)
+	if err != nil {
+		handleError(w, http.StatusBadRequest, err.Error())
+		return
+	}
+	resp := make(map[string]models.FavoriteQuickstart)
+	resp["data"] = favQuickstart
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(resp)
```
Comment (Hyperkid123) [inline pkg/routes/favorite_quickstart.go:114]: Just a note here for a follow-up task. Make a generic response handler.

**Review by Hyperkid123 — COMMENTED** [review]
We should hide the new endpoints and DB tables creation/migration via env variable (don't want to onboard to feature flags now) until we have the entire workflow working. We are still missing the migration which should be done in a follow-up.

File: pkg/routes/favorite_quickstart_test.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+func setupFavoriteQuickstartRouter() *chi.Mux {
+	r := chi.NewRouter()
+	r.Get("/", GetAllFavoriteQuickstarts)
+	r.Post("/", UpdateFavoriteQuickstart)
+	return r
+}
+
+func TestFavoriteQuickstarts(t *testing.T) {
```
Comment (BlakeHolifield) [inline pkg/routes/favorite_quickstart_test.go:44]: These look great. I think we should add some testing for error conditions to make sure the `handleError` method is tested

File: pkg/routes/favorite_quickstart.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+type FavQuickstartPayload struct {
+	QuickstartName string `json:"quickstartName"`
+	Favorite       bool   `json:"favorite"`
+}
+
+func UpdateFavoriteQuickstart(w http.ResponseWriter, r *http.Request) {
+	var accountId = r.URL.Query().Get("account")
+	if accountId == "" {
```
Comment (BlakeHolifield) [inline pkg/routes/favorite_quickstart.go:85]: This should also use the `handleError` method as above

File: pkg/routes/favorite_quickstart.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+		w.Header().Set("Content-Type", "application/json")
+		resp := make(map[string]string)
+
+		resp["msg"] = "account query param is required"
+		json.NewEncoder(w).Encode(resp)
+		return
+	}
+
+	var payload FavQuickstartPayload
+	if err := json.NewDecoder(r.Body).Decode(&payload); err != nil {
```
Comment (BlakeHolifield) [inline pkg/routes/favorite_quickstart.go:96]: `handleError` again

**Review by Hyperkid123 — APPROVED** [review]

**Pull request merged**